File: artifacts/api-server/src/lib/providers/qos.ts
import { envPositiveNumber } from "../env";

/**
 * Per-provider quality-of-service tracking for train status upstreams.
 *
 * Every provider call made through `fetchProviderStatus` is recorded here
 * (outcome, latency, timeout) so each upstream can be watched independently
 * and the orchestrator can skip providers that keep failing instead of
 * burning a timeout on every request.
 */
export type ProviderQosOutcome = "success" | "not_found" | "upstream_error";

export interface QosRecord {
  outcome: ProviderQosOutcome;
  latencyMs: number;
  /** Only meaningful for `upstream_error`; distinguishes timeouts. */
  timeout?: boolean;
  /** Last upstream error message, for diagnostics. */
  error?: string | null;
}

export interface QosRegistryOptions {
  /**
   * Consecutive upstream failures after which a provider is treated as down
   * and skipped by the orchestrator (until `cooldownMs` elapses).
   */
  failureThreshold: number;
  /** How long a down provider is skipped before being tried again. */
  cooldownMs: number;
  /** Rolling latency window kept per provider (drives avg/p95). */
  maxLatencySamples: number;
}

export type ProviderQosStatus = "ok" | "degraded" | "down";

export interface ProviderQosSnapshot {
  name: string;
  requests: number;
  successes: number;
  notFound: number;
  upstreamErrors: number;
  timeouts: number;
  consecutiveFailures: number;
  errorRate: number;
  avgLatencyMs: number;
  p95LatencyMs: number | null;
  status: ProviderQosStatus;
  /** Whether the provider is currently being consulted by the orchestrator. */
  available: boolean;
  lastSuccessAt: string | null;
  lastError: string | null;
  lastErrorAt: string | null;
}

interface ProviderQosData {
  requests: number;
  successes: number;
  notFound: number;
  upstreamErrors: number;
  timeouts: number;
  consecutiveFailures: number;
  lastError: string | null;
  lastErrorAt: string | null;
  lastSuccessAt: string | null;
  latencies: number[];
}

export const DEFAULT_QOS_OPTIONS: Required<QosRegistryOptions> = {
  failureThreshold: 3,
  cooldownMs: 60_000,
  maxLatencySamples: 100,
};

function emptyData(): ProviderQosData {
  return {
    requests: 0,
    successes: 0,
    notFound: 0,
    upstreamErrors: 0,
    timeouts: 0,
    consecutiveFailures: 0,
    lastError: null,
    lastErrorAt: null,
    lastSuccessAt: null,
    latencies: [],
  };
}

function percentile(sorted: readonly number[], p: number): number | null {
  if (sorted.length === 0) return null;
  const index = Math.min(
    sorted.length - 1,
    Math.max(0, Math.ceil(p * sorted.length) - 1),
  );
  return sorted[index];
}

export class QosRegistry {
  private readonly options: Required<QosRegistryOptions>;
  private readonly data = new Map<string, ProviderQosData>();

  constructor(options: Partial<QosRegistryOptions> = {}) {
    this.options = { ...DEFAULT_QOS_OPTIONS, ...options };
  }

  record(name: string, event: QosRecord): void {
    const entry = this.data.get(name) ?? emptyData();
    entry.requests += 1;
    entry.latencies.push(event.latencyMs);
    if (entry.latencies.length > this.options.maxLatencySamples) {
      entry.latencies.shift();
    }

    switch (event.outcome) {
      case "success":
        entry.successes += 1;
        entry.consecutiveFailures = 0;
        entry.lastSuccessAt = new Date().toISOString();
        break;
      case "not_found":
        // An authoritative "no data" answer is a healthy response.
        entry.notFound += 1;
        entry.consecutiveFailures = 0;
        break;
      case "upstream_error":
        entry.upstreamErrors += 1;
        entry.consecutiveFailures += 1;
        entry.lastError = event.error ?? "upstream error";
        entry.lastErrorAt = new Date().toISOString();
        if (event.timeout) entry.timeouts += 1;
        break;
    }

    this.data.set(name, entry);
  }

  /**
   * True when the provider may be consulted. Providers with fewer than
   * `failureThreshold` consecutive failures — or past their cooldown — are
   * always available; unknown providers are assumed healthy.
   */
  isAvailable(name: string, now = Date.now()): boolean {
    const entry = this.data.get(name);
    if (!entry) return true;
    if (entry.consecutiveFailures < this.options.failureThreshold) return true;
    if (entry.lastErrorAt === null) return true;
    const lastErrorMs = Date.parse(entry.lastErrorAt);
    return now - lastErrorMs >= this.options.cooldownMs;
  }

  snapshot(name: string, now = Date.now()): ProviderQosSnapshot {
    const entry = this.data.get(name) ?? emptyData();
    const total = entry.requests;
    const errorRate = total === 0 ? 0 : entry.upstreamErrors / total;
    const sorted = [...entry.latencies].sort((a, b) => a - b);
    const sum = entry.latencies.reduce((acc, value) => acc + value, 0);

    let status: ProviderQosStatus = "ok";
    if (entry.consecutiveFailures >= this.options.failureThreshold) {
      status = "down";
    } else if (errorRate > 0.1 || entry.upstreamErrors > 0) {
      status = "degraded";
    }

    return {
      name,
      requests: entry.requests,
      successes: entry.successes,
      notFound: entry.notFound,
      upstreamErrors: entry.upstreamErrors,
      timeouts: entry.timeouts,
      consecutiveFailures: entry.consecutiveFailures,
      errorRate,
      avgLatencyMs: entry.latencies.length === 0 ? 0 : sum / entry.latencies.length,
      p95LatencyMs: percentile(sorted, 0.95),
      status,
      available: this.isAvailable(name, now),
      lastSuccessAt: entry.lastSuccessAt,
      lastError: entry.lastError,
      lastErrorAt: entry.lastErrorAt,
    };
  }

  /** Snapshots for the given names, in the given order, skipping unknowns. */
  snapshotFor(names: readonly string[], now = Date.now()): ProviderQosSnapshot[] {
    return names.map((name) => this.snapshot(name, now));
  }

  snapshotAll(now = Date.now()): ProviderQosSnapshot[] {
    return [...this.data.keys()].map((name) => this.snapshot(name, now));
  }

  reset(): void {
    this.data.clear();
  }
}

/** Build a registry from env, so tuning is deployable without code changes. */
export function createQosRegistryFromEnv(
  env: NodeJS.ProcessEnv,
): QosRegistry {
  return new QosRegistry({
    failureThreshold: envPositiveNumber(
      env,
      "TRAIN_STATUS_QOS_FAILURE_THRESHOLD",
      DEFAULT_QOS_OPTIONS.failureThreshold,
    ),
    cooldownMs: envPositiveNumber(
      env,
      "TRAIN_STATUS_QOS_COOLDOWN_MS",
      DEFAULT_QOS_OPTIONS.cooldownMs,
    ),
    maxLatencySamples: envPositiveNumber(
      env,
      "TRAIN_STATUS_QOS_LATENCY_SAMPLES",
      DEFAULT_QOS_OPTIONS.maxLatencySamples,
    ),
  });
}

/** Shared process-wide registry; tests use their own instance where needed. */
export const defaultQosRegistry = createQosRegistryFromEnv(process.env);

File: rust/crates/qos/tests/qos.rs
//! Port of `lib/providers/qos.test.ts` — every case 1:1.

use tt_qos::{
    create_qos_registry_from_env, QosOptions, QosOutcome, QosRecord, QosRegistry, QosStatus,
    DEFAULT_QOS_OPTIONS,
};

const NOW: i64 = 1_700_000_000_000;

fn error_event(latency_ms: f64) -> QosRecord {
    QosRecord {
        outcome: QosOutcome::UpstreamError,
        latency_ms,
        timeout: false,
        error: None,
    }
}

#[test]
fn starts_empty_and_reports_healthy_for_unknown_providers() {
    let qos = QosRegistry::default();
    assert!(qos.is_available("paytm", NOW));
    let snapshot = qos.snapshot("paytm", NOW);
    assert_eq!(snapshot.requests, 0);
    assert_eq!(snapshot.status, QosStatus::Ok);
    assert!(snapshot.available);
    assert_eq!(snapshot.error_rate, 0.0);
}

#[test]
fn tracks_successes_and_resets_consecutive_failures() {
    let qos = QosRegistry::default();
    qos.record("a", &error_event(100.0));
    qos.record("a", &error_event(100.0));
    qos.record(
        "a",
        &QosRecord {
            outcome: QosOutcome::Success,
            latency_ms: 50.0,
            timeout: false,
            error: None,
        },
    );

    let snapshot = qos.snapshot("a", NOW);
    assert_eq!(snapshot.requests, 3);
    assert_eq!(snapshot.successes, 1);
    assert_eq!(snapshot.upstream_errors, 2);
    assert_eq!(snapshot.consecutive_failures, 0);
    assert!(snapshot.last_success_at.is_some());
}

#[test]
fn treats_not_found_as_a_healthy_authoritative_answer() {
    let qos = QosRegistry::default();
    qos.record(
        "a",
        &QosRecord {
            outcome: QosOutcome::NotFound,
            latency_ms: 200.0,
            timeout: false,
            error: None,
        },
    );

    let snapshot = qos.snapshot("a", NOW);
    assert_eq!(snapshot.not_found, 1);
    assert_eq!(snapshot.consecutive_failures, 0);
    assert!(snapshot.available);
    assert_eq!(snapshot.status, QosStatus::Ok);
}

#[test]
fn computes_error_rate_and_degrades_status_on_repeated_failures() {
    let qos = QosRegistry::default();
    qos.record("a", &error_event(10.0));
    qos.record("a", &error_event(10.0));
    qos.record(
        "a",
        &QosRecord {
            outcome: QosOutcome::Success,
            latency_ms: 10.0,
            timeout: false,
            error: None,
        },
    );

    let snapshot = qos.snapshot("a", NOW);
    assert!((snapshot.error_rate - 2.0 / 3.0).abs() < 1e-9);
    assert_eq!(snapshot.status, QosStatus::Degraded);
}

#[test]
fn marks_a_provider_down_after_the_failure_threshold() {
    let qos = QosRegistry::new(QosOptions {
        failure_threshold: 3,
        cooldown_ms: 60_000,
        max_latency_samples: 10,
    });

    qos.record("a", &error_event(10.0));
    qos.record("a", &error_event(10.0));
    assert!(qos.is_available("a", NOW));

    qos.record("a", &error_event(10.0));
    assert!(!qos.is_available("a", NOW));
    assert_eq!(qos.snapshot("a", NOW).status, QosStatus::Down);
}

#[test]
fn reopens_a_provider_after_the_cooldown_elapses() {
    let qos = QosRegistry::new(QosOptions {
        failure_threshold: 1,
        cooldown_ms: 60_000,
        max_latency_samples: 10,
    });

    qos.record("a", &error_event(10.0));
    let last_error_at = qos.snapshot("a", NOW).last_error_at.expect("recorded");
    let failed_at = chrono::DateTime::parse_from_rfc3339(&last_error_at)
        .expect("ISO-8601")
        .timestamp_millis();

    assert!(!qos.is_available("a", failed_at));
    assert!(qos.is_available("a", failed_at + 60_000));
}

#[test]
fn computes_avg_and_p95_latency_from_the_rolling_window() {
    let qos = QosRegistry::default();
    for i in 1..=20 {
        qos.record(
            "a",
            &QosRecord {
                outcome: QosOutcome::Success,
                latency_ms: f64::from(i),
                timeout: false,
                error: None,
            },
        );
    }

    let snapshot = qos.snapshot("a", NOW);
    assert!((snapshot.avg_latency_ms - 10.5).abs() < 1e-9);
    assert_eq!(snapshot.p95_latency_ms, Some(19.0));
}

#[test]
fn caps_the_latency_window_at_max_latency_samples() {
    let qos = QosRegistry::new(QosOptions {
        max_latency_samples: 3,
        ..QosOptions::default()
    });
    for i in 1..=5 {
        qos.record(
            "a",
            &QosRecord {
                outcome: QosOutcome::Success,
                latency_ms: f64::from(i),
                timeout: false,
                error: None,
            },
        );
    }

    let snapshot = qos.snapshot("a", NOW);
    assert!((snapshot.avg_latency_ms - 4.0).abs() < 1e-9);
    assert_eq!(snapshot.p95_latency_ms, Some(5.0));
}

#[test]
fn counts_timeouts_separately() {
    let qos = QosRegistry::default();
    qos.record(
        "a",
        &QosRecord {
            outcome: QosOutcome::UpstreamError,
            latency_ms: 10_000.0,
            timeout: true,
            error: None,
        },
    );

    let snapshot = qos.snapshot("a", NOW);
    assert_eq!(snapshot.timeouts, 1);
    assert_eq!(snapshot.upstream_errors, 1);
}

#[test]
fn snapshots_for_a_list_of_names_in_order() {
    let qos = QosRegistry::default();
    qos.record(
        "b",
        &QosRecord {
            outcome: QosOutcome::Success,
            latency_ms: 1.0,
            timeout: false,
            error: None,
        },
    );
    qos.record(
        "a",
        &QosRecord {
            outcome: QosOutcome::Success,
            latency_ms: 2.0,
            timeout: false,
            error: None,
        },
    );

    let snapshots = qos.snapshot_for(&["a", "b", "c"], NOW);
    let names: Vec<&str> = snapshots.iter().map(|s| s.name.as_str()).collect();
    assert_eq!(names, vec!["a", "b", "c"]);
    assert_eq!(snapshots[2].requests, 0);
}

#[test]
fn reset_clears_all_recorded_state() {
    let qos = QosRegistry::default();
    qos.record("a", &error_event(10.0));
    assert_eq!(qos.snapshot_all(NOW).len(), 1);

    qos.reset();
    assert_eq!(qos.snapshot_all(NOW).len(), 0);
    assert!(qos.is_available("a", NOW));
}

#[test]
fn defaults_match_the_typescript_defaults() {
    assert_eq!(DEFAULT_QOS_OPTIONS.failure_threshold, 3);
    assert_eq!(DEFAULT_QOS_OPTIONS.cooldown_ms, 60_000);
    assert_eq!(DEFAULT_QOS_OPTIONS.max_latency_samples, 100);
    assert_eq!(QosOptions::default(), DEFAULT_QOS_OPTIONS);
}

#[test]
fn create_from_env_reads_process_environment_and_fails_open() {
    std::env::set_var("TRAIN_STATUS_QOS_FAILURE_THRESHOLD", "5");
    std::env::set_var("TRAIN_STATUS_QOS_COOLDOWN_MS", "120000");
    std::env::set_var("TRAIN_STATUS_QOS_LATENCY_SAMPLES", "50");
    let registry = create_qos_registry_from_env();
    std::env::remove_var("TRAIN_STATUS_QOS_FAILURE_THRESHOLD");
    std::env::remove_var("TRAIN_STATUS_QOS_COOLDOWN_MS");
    std::env::remove_var("TRAIN_STATUS_QOS_LATENCY_SAMPLES");

    for _ in 0..4 {
        registry.record("a", &error_event(10.0));
    }
    assert!(registry.is_available("a", NOW), "threshold is 5, not 3");
    registry.record("a", &error_event(10.0));
    assert!(!registry.is_available("a", NOW));
}
